alts/src/main/java/io/grpc/alts/AltsChannelBuilder.java

```diff
@@ -148,29 +141,30 @@ AltsProtocolNegotiator getProtocolNegotiatorForTest() {
 
   @VisibleForTesting
   @Nullable
-  AltsClientOptions getAltsClientOptionsForTest() {
-    return handshakerOptionsForTest;
+  AltsClientOptions.Builder getAltsClientOptionsBuilderForTest() {
+    return handshakerOptionsBuilder;
   }
 
   private final class ProtocolNegotiatorFactory
       implements InternalNettyChannelBuilder.ProtocolNegotiatorFactory {
+
     @Override
     public AltsProtocolNegotiator buildProtocolNegotiator() {
-      final AltsClientOptions handshakerOptions = handshakerOptionsBuilder.build();
       TsiHandshakerFactory altsHandshakerFactory =
           new TsiHandshakerFactory() {
             @Override
-            public TsiHandshaker newHandshaker() {
+            public TsiHandshaker newHandshaker(String authority) {
               // Used the shared grpc channel to connecting to the ALTS handshaker service.
               // TODO: Release the channel if it is not used.
               // https://github.com/grpc/grpc-java/issues/4755.
+              AltsClientOptions.Builder clientOptionsBuilder = handshakerOptionsBuilder;
```

Comment:
This is not safe, as the options can continue mutating after `build()` returns. It's not clear to me why the `handshakerOptionsBuilder.build()` above was removed.

Reply:
The reason is that AltsClientOption is immutable once built. The authority information is per ProtocolNegotiator (sub-channel), but handshakerOptionBuilder is per channel. We need to copy handshakerOption. 

Alternative is allow setTargetName on AltsClientOption, which makes it mutable. Build AltsClientOptions first, then in newHandshaker(), make a copy, set authority. Is this something you preferred?